earth_enterprise/src/fusion/autoingest/sysman/AssetVersionD.cpp

```diff
@@ -18,11 +18,118 @@
 #include <AssetThrowPolicy.h>
 #include <khFileUtils.h>
 #include <khstl.h>
+#include <algorithm>
 #include "AssetVersionD.h"
 #include "AssetD.h"
 #include "Asset.h"
 #include "Misc.h"
 #include "khAssetManager.h"
+#include "fusion/autoingest/MiscConfig.h"
+
+// ****************************************************************************
+// ***  StateUpdateNotifier
+// ****************************************************************************
+std::shared_ptr<AssetVersionImplD::StateChangeNotifier>
+AssetVersionImplD::StateChangeNotifier::GetNotifier(
+    AssetVersionImplD * const assetVersion,
+    std::shared_ptr<StateChangeNotifier> callerNotifier) {
+  if (callerNotifier) {
```

Comment:
I think you could even turn this into a singleton, and use one global `StateChangeNotifier` instance for all state changes.  Then, you wouldn't need to pass a `callerNotifier` instance around, either.  But, I think this would be OK for now.

Reply:
My concern about the singleton idea is that this code is written to be multi-threaded. I know the way the implementation works right now is effectively single threaded, but I don't want to bury a nasty bug in here for someone to find later.

One other point I'd like to make - it's not possible to save up all the notifications and do them at once, at least the way the code is currently written. When something is notified, it may update its state, which may cause something else to be notified. So every notification may produce more notifications. We could fix that as well, but it would be a much bigger refactor.